README.md
#MCRAM 1.0.0b1 (beta)

Dependencies: Java 7+ (Java 8 recommended)
Support Operating Systems: Linux, MacOS X, Windows

MCRAM is designed to help run a Minecraft server entirely in RAM on any operating system. It will set up the automation of starting the server 100% in memory and then saving/syncing it back to the drive occasionally. This helps to improve performance and extend the computer's drive's health and longevity.

## INSTALLATION
All: Install the Java Runtime Environment (JRE) 7 or 8.
Windows: Install "imdisk" from [here](http://www.ltr-data.se/files/imdiskinst.exe) for handling RAM disks.

If you do not have "git" installed on your local machine to clone the repository, simply download the zip file from [here](https://github.com/ekultails/mcram/archive/master.zip) instead.

## USAGE
MCRAM is bundled into a convient jar file. If you run it with no commands, it will interactively prompt you for all of the information it needs. At the bare minimum, you need to specify the source directory (a full path to the Minecraft server's jar file) and the destination directory (on Windows, an unused drive letter to create a RAM disk on).

On UNIX-like systems, the mcramctl utility should be run with either "sudo" or as the root user to allow the mounting of the RAM file system.
```
# git clone https://github.com/ekultails/mcram.git
# cd mcram/java/built/
# java -jar mcramctl.jar --help
mcramctl options
	--help, -h
		display these help options
	--source-dir, -s
		provide the full path to the Minecraft server's jar file
		Windows: folders should be seperated by "/" and not "\"
	--destination-dir, -d
		where to mount the RAM disk to
		Unix: provide an empty directory
			default: /mcram/
		Windows: provide an unused drive letter
			default: M:/
	--run-ram, -rr
		specify (in MB) the amount of RAM to run the Minecraft server
		default: 512
	--mount-ram, -mr
		specify (in MB) the amount of RAM to use for mounting the RAM disk
		default: 512
	--sync-time, -t
		set how long to wait (in minutes) before copying the server back to the disk
		default: 60
	--execute, -e
		send a command to the Minecraft server
	--stop
		stop the Minecraft server
	--verbose, -v
		display debugging information
	--version, -V
		show current MCRAM version

```
After a server has been started with MCRAM, commands for the Minecraft server can be run by using the "-e" or "--execute" arguements. For example, the command below will run on the server to change the user Skywalker's gamemode to creative (1).
```
$ java -jar mcramctl -e gamemode 1 Skywalker
```

## DEVELOPERS
### BRANCHES

* master = Stable. Normal end-users should use this branch.
* dev = Development. Features are added, modified, and/or removed and after amble testing these changes will be pushed to the master branch.

### SOURCE CODE
All of the source code is located in the java/ directory. Currently there are two files, "mcramctl" to handle the user-interface and "Mcramd" to control the background daemon. After making any changes you should rebuild the latest jar file.
```
$ cd java/
$ rm built/mcramctl.jar
$ jar -cfmv built/mcramctl.jar manifest.txt *.class
```
This creates (c) a new jar file (f) with the manifest file (e; this specifies, at least, what is the main class) and will verbosely show us the process of building the jar (v).

### SOCKETS
All commands are sent to the "mcramd" daemon via a socket-like text file.
```
Linux/MacOSX socket: /tmp/mcramd.sock
Windows socket: C:/Users/%USERNAME%/AppData/Local/Temp/mcramd.sock
```
Mcramd ONLY reads the first line in the socket. This means that it needs to be cleared out every time a new command is sent to it. An example below demonstrates the ability to manually run commands on a server via the socket. This essentially pushes "/say Use the Force" to standard input for the Minecraft server's console, executing the command.
```
$ echo "mcramd:exec,say Use the Force" > /tmp/mcramd.sock
```
Alternatively, server commands can be executed via the mcramctl utilty. Only the "-e" or "--execute" argument should be given and then the rest should be the full command.
```
$ java -jar mcramctl.jar --execute say Hello Imperial Senate
```
Mcramd currently recognizes 3 flags. Every flag is then seperated by a comma "," to give each of the options.

1. mcramd:start
  * All of the options for mounting and starting the Minecraft server should be listed here.
  * Example 1 - Mount 512MB of RAM onto the /tmpfs/ directory. Then sync over the server from /home/user/mc_server/ to /tmpfs/. Start the Minecraft server, giving Java 1024MB of RAM to run with.
    *  mcramd:start,/tmpfs,512,/home/user/mc_server,1024,linux,debug:false
  * Example 2 - Mount 1024MB of RAM onto the partition letter "R:" on Windows. Then sync over the server from C:/Users/Steve/server/ to R:/. Start the Minecraft server, giving Java 2048MB of RAM to run with. Debugging is enabled here to see more verbose information about what's being executed.
    *  mcramd:start,R:/,1024,C:/Users/Steve/server/,2048,windows,debug:true
2. mcramd:exec
  * Provide the full command to execute on the server.
  * Example
    * mcramd:exec,/say Hello World
3. mcramd:stop
  * No options are required. The "stop" command will be sent to the server, MCRAM will wait 1 second, and then copy over the files from RAM back to the disk.

## LIMITATIONS
These are current limitations that will be addressed in future releases.
* MCRAM currently only handles running one server at a time in RAM.
* Standard output from the Minecraft server is not displayed.
* No service utilies for automating the start of MCRAM on boot.

## DISCLAIMER
This program does NOT provide any binaries of the Minecraft game. These must be obtained by Mojang, Microsoft, and/or any other appropriate companies officially associated with the Minecraft brand. All intellectual properties, brands, trandmarks, and/or copyrights are owned by their respective owners/companies.

## License
This software is licensed under the GPLv3. More information about this can be found in the included "LICENSE" file or online at: http://www.gnu.org/licenses/gpl-3.0.en.html


CHANGELOG.md
# MCRAM Change Log
## [1.0.0b1] 
### Added
* Created a change log
* Rewritten from Bash into Java
* Windows support
* CLI options
  * Ability to send commands to the Minecraft server
  * Ability to stop a server


install.sh
#!/bin/sh
##MCRAM - Installer - install.sh
##0.5 BETA
mcramv="0.5.0"

echo -e "\e[1m \e[36m"; #Changes color to light blue 

req1="screen"; 
req2="sudo"; 

if [[ $(type $req1 2>&1 | grep -c 'not found') -eq 1 ]]; 
	then echo "Dependency $(echo $req1) not found. Please install it via your package manager."; exit
elif [[ $(type $req2 2>&1 | grep -c 'not found') -eq 1 ]]; 	
	then echo "Dependency $(echo $req2) not found. Please install it via your package manager"; exit
else echo 'All dependencies met!'; 
fi

echo 'Please provide full folder paths. For example: /home/myuser/Documents/MC/'
read -p 'Where is your server located? ' mclocation; while [[ ! -f "$mclocation/server.properties" ]]; do read -p 'Invalid location. Please try again: ' mclocation; done 
if [[ $(\ls $mclocation | grep -c jar) -eq 1 ]]; 
	then mcstart=$(\ls $mclocation | grep jar)
elif [[ $(\ls $mclocation | grep -c jar) -eq 0 ]];
	then echo "No executable server jar file was found in $mclocation";
	echo "Try re-downloading the minecaft_server.jar from minecraft.net"; exit
else 
	echo "Here is a list of possible server jar files: "; \ls $mclocation | grep jar; 
	read -p 'Which one would you like to use to start your server with? ' mcstart; 
	while [[ $(\ls $mclocation | grep -Pc "^${mcstart}$") -ne 1  ]]; 
		do read -p 'Invalid entry. Please try again: ' mcstart; 
	done
fi

read -p 'What directory do you want RAM mounted onto for MC? ' ramlocation;
sudo mkdir -p $ramlocation

availableRam=$(expr $(grep MemTotal /proc/meminfo | awk {'print $2'}) / 1024)
echo "Your computer has ${availableRam}MB of RAM"

read -p 'How much RAM (in MB) would you like to use for mounting Minecraft into RAM (default is 256)? ' mountRam;
while [[ ${mountRam} -gt ${availableRam} ]];
	do read -p "Incorrect value. Please pick a value lower than ${availableRam}MB" mountRam;
done
if [[ $(echo ${mountRam}) -eq "" ]]; 
	then mountRam="256"
fi

read -p 'What amount of RAM (in MB) would you like to use for running Minecraft? (Default is 256): ' mcstartram;
while [[ ${mcstartram} -gt ${availableRam} ]];
	do read -p "Incorrect value. Please pick a value lower than ${availableRam}MB" mcstartram;
done
if [[ $(echo ${mcstartram}) -eq "" ]]; 
	then mcstartram="256"
fi

read -p 'How long (in minutes) do you want to wait for the server to sync to the drive again? (Default is 60): ' synctime;
if [[ $(echo $synctime) -eq "" ]]; 
	then synctime="60"
fi

if [[ -f /etc/sysctl.conf ]]; then\
	echo "EXPERIMENTAL. Adjusting the swappiness settings does not work on all systems."
	echo "Your swappiness level is $(cat /proc/sys/vm/swappiness) out of 100."
        read -p "The lower the number, the less likely your system and the tmpfs is to use swap. Would you like to adjust this (Y/N)? " answer;
        if [[ "$answer" == "Y" || "$answer" == "y" ]]; then\
            read -p "What level number do you want to use? " answer;
		#Update the system configuration to use this swappiness level after reboots
		sed -i s/^vm.swappiness/\#vm.swappiness/g /etc/sysctl.conf
		echo "vm.swappiness=${answer}" >> /etc/sysctl.conf
		#Apply the new setting now
		sudo swapoff -a; sudo swapon -a
        fi
fi
	
#currentdate=$(date +%Y-%m-%d_%Hh.%Mm.%Ss)

echo "Setting up the automted MCRAM service now..."
#Setup the cron
	#The sed replace commands need an escape character for every directory slash "/" in the bash variables. The below fixes that.
mclocation_sedfix=$(echo $mclocation | sed 's/\//\\\//g')
ramlocation_sedfix=$(echo $ramlocation | sed 's/\//\\\//g')

#Create the MCRAM install directory
mkdir ~/mcram/

mccronlocation=$(echo ~/mcram/mccron.sh)
cp -af ./mccron-template.sh ~/mcram/mccron.sh
sed -i s'/${mountRam}/'"$mountRam"'/g' ~/mcram/mccron.sh
sed -i s'/$mclocation/'"$mclocation_sedfix"'/g' ~/mcram/mccron.sh
sed -i s'/$ramlocation/'"$ramlocation_sedfix"'/g' ~/mcram/mccron.sh
sed -i s'/${mcstartram}/'"$mcstartram"'/g' ~/mcram/mccron.sh
sed -i s'/$mcstart/'"$mcstart"'/g' ~/mcram/mccron.sh
sed -i s'/ ${synctime}/'" $synctime"'/g' ~/mcram/mccron.sh


if [[ $(type systemctl 2>&1 | grep -c "not found") -eq 0 ]];
	then mccronlocation_sedfix=$(echo $mccronlocation | sed 's/\//\\\//g')
	sed -i s'/$mccronlocation/'"$mccronlocation_sedfix"'/g' ./mcram.service
	chmod 750 ./mcram.service
	sudo cp -af ./mcram.service /usr/lib/systemd/system/mcram.service
	sudo systemctl enable mcram; sudo systemctl start mcram &
	mcramJobID=$(jobs | grep "sudo systemctl start mcram" | grep -oP "[0-9][0-9]*")
	disown %${mcramJobID}
elif [[ $(type crontab 2>&1 | grep -c "not found") -eq 0 ]];
	then echo -e "$(crontab -l)\n@reboot /bin/sh $HOME/mcram/mccron.sh" | crontab -
else echo "systemd and crontab are not installed, please install one of these services"
fi

sudo chmod o+rw /dev/pts/2; sudo chmod 755 /run/screens; #Fixes screen issues when running as the user
chmod 750 ~/mcram/mccron.sh #Makes the cron executable

echo -e "MCRAM ${mcramv} has been installed.\nPlease report any problems or suggestions to https://github.com/ekultails/mcram/"
sh ~/mcram/mccron.sh &

echo -e "\e[0;00m"; #Resets the colors

###CHANGELOG
##FUTURE RELEASE
#	Automatically install dependencies on Debian, RHEL, and Arch based hosts
#	Create/enforce strict permissions
#	Add the ability to automatically generate backups before setting up the server in RAM
#
#=-=-=CHANGES SINCE LAST MAJOR RELEASE=-=-=#
##0.4 Beta
#	Implemented:
#		Mac OSX (Darwin) support
#	Bug fix:
#		mounts are now done manually to avoid /etc/fstab issues
#		
##0.3
#	Implemented:
#		Uninstaller created
#		Amount of mounted RAM can is now determined by megabytes instead of a percentage of available RAM
#	Bug Fix:
#		rsync wasn't always syncing properly; Switched arguments for rsync from -varP to -varuP
#		systemd service installation no longer hangs; the mcram.service is now started detatched in the background
#		Do not allow memory usage higher than the available systems RAM
#
##0.2
#	Implemented:
#		"sudo" is now used for non-root sudo users to install/set-up MCRAM
#		Added a systemd start-up script "mcram.service"
#		systemd timer unit is used instead of the crontab if systemd is available
#		Sync time intervals can now be changed
#		MCRAM now installs itself to the user's home directory in a folder called "mcram"
#		"replace" has been replaced with "sed" commands for better compatibility across platforms
#
##0.1 Alpha
#	Implemented:
#		Sets up a cron for the Minecraft server to start on boot
#		Mounts MC server into memory/RAM
#		Syncs the server from RAM every hour
#




uninstall.sh
#!/bin/sh
##MCRAM - Uninstaller - uninstall.sh

echo "MCRAM is now being uninstalled..."

#Send commands to the screens to quit the sessions
mcrsyncScreen=$(sudo screen -ls | grep -c "mcrsync")
if [[ ${mcrsyncScreen} > 1 ]]; 
	then echo "You have more than one 'mcrsync' screen. Please terminate these manually."
elif [[ ${mcrsyncScreen} == 0 ]];
	then echo "There is no 'mcrsync' screen running."
else screen -S mcrsync -X quit	
fi

mcramScreen=$(sudo screen -ls | grep -c "mcram")
if [[ ${mcramScreen} > 1 ]]; 
	then echo "You have more than one 'mcram' screen. Please terminate these manually."
elif [[ ${mcramScreen} == 0 ]];
	then echo "There is no 'mcram' screen running."
else\
	screen -p 0 -S mcram -X eval 'stuff "save-all"\015';
	sleep 5s;
	screen -p 0 -S mcram -X eval 'stuff "stop"\015';
	screen -S mcram -X quit
fi

if [ -f ~/mcram/etc-fstab ];
	then "Attempting to remove the RAM mounts..."
	echo "Here are the differences between your original and current /etc/fstab" 
	diff -y /etc/fstab ~/mcram/etc-fstab
	read -p "Would you like to restore the original /etc/fstab (Y/N)? " restorefstab;
	if [[ ${restorefstab} == "Y" || ${restorefstab} == "y" ]];
		then currentdate=$(date +%Y-%m-%d_%Hh.%Mm.%Ss)
		sudo mv -v /etc/fstab /etc/fstab${currentdate}
		sudo cp -v ~/mcram/etc-fstab /etc/fstab
		mount -a
	fi
fi

#Delete MCRAM files
echo "Removing all MCRAM files..."
rm -rf ~/mcram/ /usr/lib/systemd/system/mcram.service > /dev/null 2>&1

#Remove cron if it exists
if [[ $(type crontab 2>&1| grep -c "not found") == 0 ]];
	then echo "Removing any possible crons that were set..." 
	crontab -l | grep -v "/mcram/" | crontab -
fi

echo "MCRAM has now been uninstalled."


java/mcramd.java
import java.io.*;
import java.io.IOException; // I/O error exception handling
import java.io.File;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.StandardCopyOption; // file copying
import java.util.Arrays;
import java.util.concurrent.TimeUnit; // TimeUnit sleep functions resides here
import java.util.Scanner;

class mcramd {

	public static String readSock(String fileName) {
	
		File fileToRead = new File(fileName);
		FileReader fileReader = null;
		
		try {
			fileReader = new FileReader(fileToRead);
		} catch (IOException e) {
			e.printStackTrace();
		}
			
		BufferedReader fileBufferReader = new BufferedReader(fileReader);
		String returnSocket = null; // initlalize the variable before
									// we enter the try/catch statement
	
		try {
			returnSocket = fileBufferReader.readLine();
		} catch (IOException e) {
			e.printStackTrace();
		}

		return returnSocket;
		
	}

    public static void mountRAMMac() {
    
        System.out.println("Stub");
        
    }

    public static void mountRAMLinux(String mountram, String destinationdir) {
        
        String mountCmd = ("mount -t tmpfs -o defaults,noatime,size=" +
                          mountram + "M tmpfs " + destinationdir);
        System.out.println(mountCmd);
        File runDir = new File("/tmp");
        execCmd(mountCmd, runDir);
        
    }
    
    public static void mountRAMWindows() {
    
        System.out.println("Stub");
        
    }

    // return the contents of a directory
    public static File[] dir(String dirname) {

        // initate a File-type variable that can be used to get the contents of a directory
        File directory = new File(dirname);

        File[] listOfFiles = directory.listFiles();

        for (File file: listOfFiles){
            System.out.println(file.getName());
        }

        return listOfFiles;

    }

    // copy files
    public static void cp(String dst_dir, File[] listOfFiles) throws IOException  {

        for(File filename : listOfFiles) {
            Path src_file = filename.toPath();
            Path dst_file = new File(dst_dir + filename.getName()).toPath();
            Files.copy(src_file, dst_file, StandardCopyOption.REPLACE_EXISTING);
        }

    }

    public static void stdin(Process p, String command) {
 
        // a Scanner is used to push standard input to the process   
        Scanner stdin_scanner = new Scanner(System.in);
        PrintWriter stdin = new PrintWriter(p.getOutputStream());
        stdin.println(command);
        stdin.flush();  
        
    }


    public static Process execCmd(String runCmd, File runDir) {
    
        Process p = null; /* a variable must be defined outside of the 
                             try/catch scope and be assigned a value of at least null */
    
        try {
            // we are not using any special environment variables
            String defaultEnv[] = {""}; 
            p = Runtime.getRuntime().exec(runCmd, defaultEnv, runDir);
        } catch(IOException e) {
            e.printStackTrace();
        }
        
        return p;
        
    }

    public static void main(String[] args) {

        while (true) {
            
            // initalize the object from our class
            mcramd mcramd_obj = new mcramd();
            
            String mountram = "512";
            String destinationdir = "/ram";
            mcramd_obj.mountRAMLinux(mountram, destinationdir);
            
            String sockContent = mcramd_obj.readSock("/tmp/mcramd.sock");
            String[] sockSplit = sockContent.split(",");
			for (String value : sockSplit) {
				System.out.println(value);
			}
			
            try {
                TimeUnit.SECONDS.sleep(10);
            } catch(InterruptedException e) {
                e.printStackTrace();
            }
            
            /* example copy below:
            String dirname = "/tmp/tmp1/";
            File[] listOfFiles = dir(dirname);
            String dir2 = "/tmp/tmp2/";
            try {
                cp(dir2, listOfFiles);
            } catch(IOException e){
                e.printStackTrace();
            }
            */

            // this will run a sample command on the Minecraft server 
            // to print out "hello" to all players
            String command = "say hello"; // DELME
            String java_prefix = System.getProperty("java.home");
            // Minecraft needs to be run from the directory that the eula.txt is in
            // this is where the minecraft_server*.jar should exist
            File runDir = new File("C:/path/to/minecraft/server/");
            String runCmd = (java_prefix + "/bin/java" + 
                             " -jar C:/path/to/minecraft/server/minecraft_server.1.9.jar");
            Process mc_server = mcramd_obj.execCmd(runCmd, runDir);
            
            int count = 1;

            while (count < 6) {

                try {
                    TimeUnit.SECONDS.sleep(10);
                } catch(InterruptedException e) {
                    e.printStackTrace();
                }
                
                mcramd_obj.stdin(mc_server, command); 
                count++;
            }
            
            command = "stop"; // finally, stop the server after testing
            mcramd_obj.stdin(mc_server, command);   
                
        }
 
   }

}


mccron-template.sh
#!/bin/sh
#MCRAM - Cron - mccron.sh

if [[ `uname -s` -eq "Darwin" ]];
        then echo "Mac"
        # size = INPUT_IN_MB * 1024 * 1024 / 512
        size=`expr ${mountRam} \* 1024 \* 1024 / 512`
        tmpPart=`hdid -nomount ram://${size} | grep '/dev/disk'`
        newfs_hfs ${tmpPart}
        mkdir -p $ramlocation
        mount -t hfs ${tmpPart} $ramlocation
else mkdir -p $ramlocation;\
        mount -t tmpfs -o defaults,noatime,size=${mountRam}MB tmpfs $ramlocation;
fi

if [[ ! $(ps aux | grep "$mcstart" | grep java | grep -v grep) ]]
	then #start server
		screen -dmS mcram sh; #creates a screen "mc" running bash for the server
		screen -dmS mcrsync sh; #creates a screen for rsyncing the data
		screen -S mcram -p 0 -X stuff $'rsync -avur $mclocation/ $ramlocation/\n' #rsync into RAM
		screen -S mcram -p 0 -X stuff $'cd $ramlocation/\n' #this allows the server to properly load the eula.txt license agreement
		screen -S mcram -p 0 -X stuff $'java -Xmx${mcstartram}M -Xms${mcstartram}M -jar ./$mcstart nogui\n' #start the server
fi

while [[ 1 -gt 0 ]]; do	
	sleep ${synctime}m;
	screen -p 0 -S mcram -X eval 'stuff "save-all"\015';
	sleep 60;
	screen -S mcrsync -p 0 -X stuff $'rsync -varuP $ramlocation/ $mclocation/; echo "Scheduled sync complete on $(date)."\n' #rsync the data back to the drive
done




java/mcramctl.java
import java.io.File; // file and folder handling
import java.io.IOException; // I/O error exception handling
import java.io.PrintWriter;
import java.util.Scanner; // stdin

class mcramctl 
{
    
    public static String socketText = null;
    // this is only turned on if "verbose" debugging mode is enabled
    public static boolean debugMode = false;

    public static void debug(String msg) 
    {
        if (debugMode == true) 
        {
            System.out.println("DEBUG: " + msg);
        }
    }

    // find the current operating system
    public static String findOS() 
    {   
        String osName = System.getProperty("os.name");
        String shortOSName = null;
        
        if (osName.toLowerCase().contains("mac")) 
        {
            shortOSName = "mac";
        } else if (osName.toLowerCase().contains("linux")) 
        {
            shortOSName = "linux";
        } else if (osName.toLowerCase().contains("windows")) 
        {
            shortOSName = "windows";
        } else 
        {
            System.out.println("Unsupported operating system. Exiting...");
            System.exit(1);
        }   
        debug("os=" + shortOSName);
        return shortOSName;
    }

    public static void writeToFile(String fileName, String text) 
    {
        try 
        {
            PrintWriter editor = new PrintWriter(fileName, "UTF-8");
            editor.println(text);
            editor.close();
        } catch (IOException e) 
        {
            e.printStackTrace();
        }   
    }

    public static String[] interactiveMode() 
    {
        Scanner stdin = new Scanner(System.in); // standard input from the user
        // get source directory
        System.out.println("What folder is your Minecraft server located in?");
        String sourceFolder = stdin.next(); // convert from Scanner to String
        // get destination directory
        System.out.println("What folder would you like use for mounting the RAM disk?"); 
        String destinationFolder = stdin.next();
        // get size for RAM disk
        System.out.println("How much RAM (in MB) do you want to use for mounting Minecraft into RAM?");
        String MBofMount = stdin.next();
        // get the amount of RAM that Java should use for running the server
        System.out.println("How much RAM (in MB) do you want to use for running Minecraft?");
        String MBofRun = stdin.next();
        // sync time
        System.out.println("How long (in minutes) to wait before syncing back to the disk?");
        String syncTime = stdin.next();

        String[] interactiveAnswers = {sourceFolder, destinationFolder, 
                                       MBofMount, MBofRun, syncTime};    
        return interactiveAnswers;
    }

    public static void help() 
    {
        System.out.printf("mcramctl options\n" + 
                          "\t--help, -h\n" +
                              "\t\tdisplay these help options\n" +
                          "\t--source-dir, -s\n" + 
                              "\t\tprovide the full path to the Minecraft server\'s jar file\n" +
                              "\t\tWindows: folders should be seperated by \"/\" and not \"\\\"\n" + 
                          "\t--destination-dir, -d\n" + 
                              "\t\twhere to mount the RAM disk to\n" +
                              "\t\tUnix: provide an empty directory\n" +
                              "\t\t\tdefault: /mcram/\n" +
                              "\t\tWindows: provide an unused drive letter\n" +
                              "\t\t\tdefault: M:/\n" + 
                          "\t--run-ram, -rr\n" +
                              "\t\tspecify (in MB) the amount of RAM to run the Minecraft server\n" +
                              "\t\tdefault: 512\n" +
                          "\t--mount-ram, -mr\n" + 
                              "\t\tspecify (in MB) the amount of RAM to use for mounting the RAM disk\n" +
                              "\t\tdefault: 512\n" +
                          "\t--sync-time, -t\n" + 
                              "\t\tset how long to wait (in minutes) before copying the server back to the disk\n" +
                              "\t\tdefault: 60\n" +
                          "\t--execute, -e\n" +
						      "\t\tsend a command to the Minecraft server\n" +
						  "\t--stop\n" +
						      "\t\tstop the Minecraft server\n" +
                          "\t--verbose, -v\n" +
                              "\t\tdisplay debugging information\n" +
                          "\t--version, -V\n" +
                              "\t\tshow current MCRAM version\n");
    }

    public static void main(String args[]) 
    {
        mcramctl mcramctl = new mcramctl(); // create an object      
        // initiate variables that will be needed later
        String sourceDir = null;
        String destinationDir = null;
        boolean isExec = false;
        boolean isStop = false;
        String cmd = "";
        String runRAM = "512";
        String mountRAM = "512";
        String syncTime = "60";
        String shortOSName = mcramctl.findOS();
        String socketFile = null;
        String[] startCheck = null;

        // only run the interactive mode if 
        // no command line arguments are given
        if (args.length == 0) 
        {
            String interactiveAnswers[] = mcramctl.interactiveMode();
            // seperate our command string by commas "," for 
            // mcramd to process through
            socketText = ("mcramd:start" + "," + interactiveAnswers[1] +
                    "," + interactiveAnswers[2] + "," + interactiveAnswers[0] + "," +
                    interactiveAnswers[3] + "," + "syncTime:" +
                    interactiveAnswers[4] + "," + shortOSName);
        } else 
        {
            // sort through the command line arguments
            for (int counter = 0; counter < args.length; counter++) 
            {
                switch (args[counter]) 
                {
                    case "--destination-dir":
                    case "-d":
                        destinationDir = args[counter + 1];
                        counter++;
                        break;
                    case "--execute":
                    case "-e":
                    	isExec = true;
						cmd = "/";
						for (int execCounter = counter + 1; execCounter < args.length; execCounter++)
						{
							cmd = cmd + args[execCounter] + " ";
						}
						counter = args.length;
                        break;
                    case "--help":
                    case "-h":
                        mcramctl.help();
                        System.exit(0);
                    case "--mount-ram":
                    case "-mr":
                        mountRAM = args[counter + 1];
                        counter++;
                        break;
                    case "--run-ram":
                    case "-rr":
                        runRAM = args[counter + 1];
                        counter++;
                        break;
                    case "--source-dir":
                    case "-s":
                        sourceDir = args[counter + 1];
                        counter++;
                        break;
                    case "--stop":
                    	isStop = true;
						cmd = "mcramd:stop"; 
						break;
                    case "--sync-time":
                    case "-t":
                        syncTime = args[counter + 1];
                        counter++;
                        break;
                    case "--verbose":
                    case "-v":
                        debugMode = true;
                        break;
                    case "--version":  
                    case "-V":
                        System.out.println("MCRAM version: 1.0.0b1");
                        System.exit(0);
                    default:
                        break;
                }
            }

            // default options
            if (destinationDir == null) 
            {
                if (shortOSName.equals("linux") || shortOSName.equals("mac")) 
                {
                    destinationDir = "/mcram/";
                } else if (shortOSName.equals("windows")) 
                {
                    destinationDir = "M:/";
                }
            }
            
            // "mcramd:start" = start the server
            // "mcramd:stop" = stop the server
            // "mcramd:exec" = exec a command on the server
				// Example: mcramd:exec,say Hello Minecraft World!
            
            if (isExec == false && isStop == false) 
            {
                // seperate our command string by commas "," for 
                // mcramd to process through
                socketText = ("mcramd:start" + "," + destinationDir +
                        "," + mountRAM + "," + sourceDir + "," +
                        runRAM + "," + "syncTime:" +
                        syncTime + "," + shortOSName + ",debug:" + debugMode);
                // "mcramd:start" = start Minecraft server
                // full mcramd:start options:
                // (1) <mount_directory>, (2) <tmpfs_size_in_MB>, 
                // (3) <source_directory>, (4) <java_RAM_exec_size_in_MB>,
                // (5) <sync_time>, (6) <operating_system_name>
                // (7) debug mode (true or false) 
                // Example: 
                // mcramd:start,/tmpfs,512,/home/user/mc_server,1024,linux,debug:false
                // OR
                // mcramd:start,R:/,1024,C:/Users/Steve/server/,1024,windows,debug:true
            } else if (isExec == true && isStop == false)
            {
                socketText = "mcramd:exec," + cmd;
            } else if (isStop == true)
            {
				socketText = "mcramd:stop";
			} 
            
            debug("socketText=" + socketText);
            
            if (socketText.contains(",null")) 
            {
                System.out.println("Required option(s) are missing " +
                                   "from string:");
                System.out.println(socketText);
                System.exit(1); 
            }
        }
        
        
        if (shortOSName.equals("linux") || shortOSName.equals("mac")) 
        {
            socketFile = "/tmp/mcramd.sock";
        } else if (shortOSName.equals("windows")) 
        {
            socketFile = "C:/Users/" + System.getProperty("user.name") +
                     "/AppData/Local/Temp/mcramd.sock";
        } else 
        {
            System.out.println("Unsupported operating system.");
            System.exit(1);
        }
        
        debug("socketFile=" + socketFile);   
        mcramctl.writeToFile(socketFile, socketText);
        // if a command is being sent to a running server then
        // do not try to start a new Minecraft server as a background thread;
        // this will cause MCRAM to "lock-up"
        if (isExec == false)
        {
			Mcramd.main(args); // execute the MCRAM daemon class
		}
    }
} 
